modules/currency.js

```diff
@@ -0,0 +1,220 @@
+
+import bidfactory from 'src/bidfactory';
+import { STATUS } from 'src/constants';
+import { ajax } from 'src/ajax';
+import * as utils from 'src/utils';
+import bidmanager from 'src/bidmanager';
+
+$$PREBID_GLOBAL$$.currency = setConfig;
+
+const DEFAULT_CURRENCY_RATE_URL = 'http://currency.prebid.org/latest.json';
+const CURRENCY_RATE_PRECISION = 4;
+
+var bidResponseQueue = [];
+var conversionCache = {};
+var currencyRatesLoaded = false;
+var adServerCurrency = 'USD';
+
+// Used as reference to the original bidmanager.addBidResponse
+var originalBidResponse;
+
+export var currencySupportEnabled = false;
+export var currencyRates = {};
+var currencyOverrides = {};
+
+export function setConfig(config) {
+  var url = DEFAULT_CURRENCY_RATE_URL;
+  if (typeof config.adServerCurrency === 'string') {
+    utils.logInfo('enabling currency support', arguments);
+
+    adServerCurrency = config.adServerCurrency;
+    if (config.conversionRateFile) {
+      utils.logInfo('currency using override conversionRateFile:', config.conversionRateFile);
+      url = config.conversionRateFile;
+    }
+    initCurrency(url);
+  } else {
+    // currency support is disabled, setting defaults
+    utils.logInfo('disabling currency support');
+    resetCurrency();
+  }
+  if (typeof config.currencyOverrides === 'object') {
+    currencyOverrides = config.currencyOverrides;
+  }
+}
+
+function initCurrency(url) {
+  conversionCache = {};
+  currencySupportEnabled = true;
+
+  if (!originalBidResponse) {
+    utils.logInfo('Installing addBidResponse decorator for currency module', arguments);
+
+    originalBidResponse = bidmanager.addBidResponse;
+    bidmanager.addBidResponse = addBidResponseDecorator(bidmanager.addBidResponse);
```

Comment:
Hi. I'm relatively new to the project and have followed this discussion with a lot of interest.
We are currently facing the issue of converting currencies as well and this what we are planning to do,
which might be of interest as well.

## Use bidderSettings

At this point using the the `pbjs.bidderSettings` to update our CPMs. If the currency module would simply be a library to call then integrating this would be a breeze. Our code looks something like this:

```typescript
export const bidderSettings: prebidjs.IBidderSettings = {
  criteo: {
    bidCpmAdjustment: (bidCpm, bid) => {
       // do the currency calculation here
       return bidCpm;
    }
  },
  appnexusAst: {
    bidCpmAdjustment: (bidCpm, bid) => {
       // do the currency calculation here
       return bidCpm;
    }
  }
};
```

If the currency module can just be imported like

```typescript
pbjs.currency.init(); // or even with a setting

const cpmInEuro = pbjs.currency.convert(cpm, 'USD', 'EUR');
```

## General approach for bid responses

Regarding a general plugin architecture. We have had very good experience with redux middlewares.
Basically you they are functions you can chain together. A `BidResponseMiddleware` would be a nice
way to abstract over all the stuff that is currently configured somewhere and applied somehow.

Also a middleware could do all the side-effects, which makes these things trivial to test, because
you can just swap them.

### Example

A middleware could have the shape of

```
(bidResponse: BidResponse) => (bidResponse: BidResponse)
```

By definition a middleware applies some side effects and calls the next
middleware. A middleware can also change the bidResponse, e.g. with currency calculations
or other stuff.

```typescript
// apply takes a list of middlewares and chains them
const defaultMiddlewares = apply(
   new AdServerTargetingMiddleware(pbjs.bidderSettings),
   new BidCpmAdjustmentMiddleware(pbjs.bidderSettings),
   new MyCustomMiddleware(),
);
```

It would also be nice to have access to the raw response from the actual SSP to work around
bugs or missing implementations. What do you think?